xds/src/main/java/io/grpc/xds/XdsResponseReader.java

```diff
@@ -0,0 +1,134 @@
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.XdsClientImpl.EndpointWatchers;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+final class XdsResponseReader implements StreamObserver<DiscoveryResponse> {
+  private static final Logger logger = XdsClientImpl.logger;
+
+  private final StreamActivityWatcher streamActivityWatcher;
+  private final EndpointWatchers endpointWatchers;
+  private final SynchronizationContext syncCtx;
+
+  XdsResponseReader(
+      StreamActivityWatcher streamActivityWatcher,
+      EndpointWatchers endpointWatchers,
+      SynchronizationContext syncCtx) {
+    this.streamActivityWatcher = streamActivityWatcher;
+    this.endpointWatchers = endpointWatchers;
+    this.syncCtx = syncCtx;
+  }
+
+  @Override
+  public void onNext(final DiscoveryResponse value) {
+
+    class HandleResponseRunnable implements Runnable {
+      @Override
+      public void run() {
+        streamActivityWatcher.responseReceived();
+
+        String typeUrl = value.getTypeUrl();
+        if (XdsClientImpl.EDS_TYPE_URL.equals(typeUrl)) {
+          handleEdsResponse(value);
+        }
+      }
+    }
+
+    syncCtx.execute(new HandleResponseRunnable());
+  }
+
+  @Override
+  public void onError(final Throwable t) {
+    logger.log(Level.INFO, "XdsClient received an error", t);
+
+    syncCtx.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            handleStreamClosed(Status.fromThrowable(t));
+          }
+        });
+  }
+
+  @Override
+  public void onCompleted() {
+    final String errMsg = "Traffic director closed the ADS stream";
+    logger.log(Level.INFO, errMsg);
+    syncCtx.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            handleStreamClosed(Status.UNAVAILABLE.withDescription(errMsg));
+          }
+        });
+  }
+
+  private void handleEdsResponse(DiscoveryResponse value) {
+    List<ClusterLoadAssignment> clusterLoadAssignments = new ArrayList<>(value.getResourcesCount());
+    try {
+      // maybe better to run this deserialization task out of syncContext?
+      for (Any any : value.getResourcesList()) {
+        clusterLoadAssignments.add(any.unpack(ClusterLoadAssignment.class));
+      }
+    } catch (InvalidProtocolBufferException | RuntimeException e) {
+      if (logger.isLoggable(Level.INFO)) {
+        logger.log(
+            Level.INFO,
+            "Unable to extract clusterLoadAssignment from EDS response: " + value,
+            e);
+      }
+      // TODO(zdapeng): NACK
+      return;
+    }
+
+    logger.log(
+        Level.FINE, "Received an EDS response: {0}", clusterLoadAssignments);
+
+    // TODO(zdapeng): ACK (but if endpointWatchers is empty then send NACK?)
```

Comment:
It seems incorrect to send a NACK when your own `endpointWatchers` is empty. What `error_detail` would you send in the NACK? It seems logical to send an ACK (new DiscoveryRequest) with empty resource names but we need to confirm it will be interpreted correctly (or find the right way to unsubscribe for resource names)